des_y6utils/piff/flag_good_regions.py

```diff
@@ -0,0 +1,249 @@
+import ngmix
+from ngmix.admom import AdmomFitter
+import galsim
+import copy
+import numpy as np
+
+ALL_BAD = 2**0
+BAD_BOX = 2**1
+
+
+def _make_a_good_box(bad_msk, verbose=False):
+    flag = 0
+
+    b = dict(
+        xmax=bad_msk.shape[1]-1,
+        xmin=0,
+        ymax=bad_msk.shape[0]-1,
+        ymin=0,
+    )
+    if not np.any(bad_msk[b["ymin"]:b["ymax"]+1, b["xmin"]:b["xmax"]+1]):
+        return b, flag
+
+    if np.all(bad_msk[b["ymin"]:b["ymax"]+1, b["xmin"]:b["xmax"]+1]):
+        flag |= ALL_BAD
+        return b, flag
+
+    def _is_ok_range(b):
+        return (
+            (b["xmin"] <= b["xmax"])
+            and (b["ymin"] <= b["ymax"])
+        )
+
+    def _condition(b):
+        return (
+            (b["xmin"] <= b["xmax"])
+            and (b["ymin"] <= b["ymax"])
+            and np.any(bad_msk[b["ymin"]:b["ymax"]+1, b["xmin"]:b["xmax"]+1])
+        )
+
+    itr = 0
+    while _condition(b) and itr < 10_000:
+        new_b = copy.deepcopy(b)
+        curr_frac = np.sum(bad_msk[b["ymin"]:b["ymax"]+1, b["xmin"]:b["xmax"]+1])
+        new_fracs = {}
+        for key, shift in zip(["ymax", "ymin", "xmax", "xmin"], [-1, 1, -1, 1]):
+            _new_b = copy.deepcopy(b)
+            _new_b[key] = b[key] + shift
+            if _is_ok_range(_new_b):
+                new_fracs[key] = np.sum(bad_msk[
+                    _new_b["ymin"]:_new_b["ymax"]+1,
+                    _new_b["xmin"]:_new_b["xmax"]+1]
+                )
+            else:
+                new_fracs[key] = np.nan
+
+        if np.all(np.isnan(list(new_fracs.values()))):
+            flag |= BAD_BOX
+            break
+
+        mval = np.nanmin(list(new_fracs.values()))
+        for key, shift in zip(["ymax", "ymin", "xmax", "xmin"], [-1, 1, -1, 1]):
+            _new_b = copy.deepcopy(b)
+            _new_b[key] = b[key] + shift
+            if new_fracs[key] == mval and _is_ok_range(_new_b):
+                new_b[key] = b[key] + shift
+                break
+
+        if verbose:
+            print("itr:", itr)
+            print("    curr # bad:", curr_frac)
+            print(
+                "    new # bad:",
+                " ".join("%s: %0.3f" % (k, v) for k, v in new_fracs.items()),
+            )
+            print("    new min val:", mval)
+            print(
+                "    edge|old|new: ",
+                " ".join("%s|%s|%s" % (k, b[k], new_b[k]) for k in b),
+            )
+        if new_b == b:
+            flag |= BAD_BOX
+            break
+
+        b = new_b
+        itr += 1
+
+    if (
+        b["xmin"] > b["xmax"]
+        or b["ymin"] > b["ymax"]
+        or np.any(bad_msk[b["ymin"]:b["ymax"]+1, b["xmin"]:b["xmax"]+1])
+    ):
+        flag |= BAD_BOX
+
+    return b, flag
```

Comment:
This algorithm doesn't necessarily find the maximum rectangle.  It just finds *a* rectangle that fits in the empty space.  So I think we can probably do better than this, although I'm not sure in practice how much area we can recover with a better algorithm.

So, I looked into this a bit, and I found this paper:

https://www.sciencedirect.com/science/article/pii/0166218X84901240

I coded up a version of their algorithm (alg1 so far -- they have a faster one I'll try next), modifying it to deal with our chunky square pixels, rather than infinitesimal points.  My implementation passes all your unit tests, so I think it's probably correct.  Then I checked a bunch of random masks with a small number of bad pixels, and it sometimes finds a much larger area than your algorithm.

Now, this may or may not make a difference in practice.  The bad masks we have aren't random.  But since we're talking about possibly losing a non-trivial amount of area, it seems worth trying to do better with this algorithm.

Reply:
More details:

Here is the script I ran comparing the two algorithms:
```
# Make masks with ~10 bad pixels to compare mer_alg1 vs Matt's algorithm.
rng = np.random.RandomState(seed=10)
for _ in range(100):
    rate = 20 / (32*64)
    bad_msk = rng.uniform(size=(32, 64)) < rate
    b1, flag1 = _make_a_good_box(bad_msk)
    assert not np.any(bad_msk[b1["ymin"]:b1["ymax"]+1, b1["xmin"]:b1["xmax"]+1])
    b2, flag2 = mer_alg1(bad_msk)
    assert not np.any(bad_msk[b2["ymin"]:b2["ymax"]+1, b2["xmin"]:b2["xmax"]+1])
    area1 = (b1['ymax']-b1['ymin']+1)*(b1['xmax']-b1['xmin']+1)
    area2 = (b2['ymax']-b2['ymin']+1)*(b2['xmax']-b2['xmin']+1)
    print(f'areas = {area1}\t{area2}')
    assert area2 >= area1
```
The output is
```
areas = 192	459
areas = 62	392
areas = 186	684
areas = 62	675
areas = 62	720
areas = 320	437
areas = 63	372
areas = 128	899
areas = 124	348
areas = 63	675
areas = 64	585
areas = 64	546
areas = 64	570
areas = 128	590
areas = 64	588
areas = 192	840
areas = 256	468
areas = 64	832
areas = 126	352
areas = 192	403
areas = 64	396
areas = 192	1180
areas = 192	658
areas = 128	460
areas = 192	648
areas = 64	320
areas = 64	323
areas = 126	728
areas = 128	392
areas = 248	351
areas = 256	609
areas = 64	424
areas = 60	360
areas = 64	351
areas = 128	486
areas = 256	475
areas = 63	462
areas = 64	768
areas = 62	360
areas = 126	378
areas = 256	529
areas = 63	465
areas = 186	456
areas = 64	351
areas = 62	440
areas = 384	486
areas = 192	520
areas = 192	434
areas = 63	490
areas = 189	494
areas = 64	319
areas = 64	468
areas = 64	468
areas = 63	513
areas = 256	392
areas = 192	480
areas = 63	392
areas = 63	377
areas = 128	468
areas = 64	408
areas = 63	290
areas = 189	650
areas = 61	637
areas = 192	336
areas = 124	574
areas = 64	464
areas = 64	455
areas = 63	391
areas = 315	713
areas = 64	464
areas = 64	442
areas = 64	480
areas = 64	646
areas = 192	319
areas = 64	522
areas = 128	437
areas = 320	512
areas = 124	440
areas = 64	966
areas = 126	651
areas = 62	527
areas = 126	484
areas = 567	576
areas = 192	665
areas = 63	440
areas = 128	288
areas = 192	513
areas = 64	640
areas = 63	363
areas = 320	690
areas = 315	338
areas = 126	392
areas = 192	702
areas = 252	512
areas = 192	572
areas = 128	810
areas = 126	410
areas = 126	375
areas = 320	528
areas = 126	473
```
So sometimes it's quite a big difference.

This algorithm is O(n^2), where n is the number of masked pixels.  I'll try to code up their faster algorithm next and if I get that working, I'll send it to you.  If not, maybe O(n^2) is tolerable.  n isn't so huge for this application.

Reply:
Thanks! I knew this was not optimal but I was hoping it was good enough for what we needed.